src/main/java/org/broadinstitute/hellbender/engine/spark/SparkSharder.java

```diff
@@ -146,48 +202,71 @@ public T apply(@Nullable Tuple2<I, Iterable<L>> input) {
+                                                                                            SAMSequenceDictionary sequenceDictionary, JavaRDD<I> intervals,
+                                                                                            int maxLocatableLength, FlatMapFunction2<Iterator<L>, Iterator<I>, T> f) {
+
         List<PartitionLocatable<SimpleInterval>> partitionReadExtents = computePartitionReadExtents(locatables, sequenceDictionary, maxLocatableLength);
+        List<SimpleInterval> firstLocatablesList = partitionReadExtents.stream().map(PartitionLocatable::getLocatable).collect(Collectors.toList());
+        Broadcast<List<SimpleInterval>> firstLocatablesBroadcast = ctx.broadcast(firstLocatablesList);
 
         // For each interval find which partition it starts and ends in.
         // An interval is processed in the partition it starts in. However, we need to make sure that
         // subsequent partitions are coalesced if needed, so for each partition p find the latest subsequent
         // partition that is needed to read all of the intervals that start in p.
-        List<Integer> maxEndPartitionIndexes = new ArrayList<>();
-        for (int i = 0; i < locatables.getNumPartitions(); i++) {
-            maxEndPartitionIndexes.add(i);
-        }
         OverlapDetector<PartitionLocatable<SimpleInterval>> overlapDetector = OverlapDetector.create(partitionReadExtents);
-        List<PartitionLocatable<I>> indexedIntervals = new ArrayList<>();
-        int lastPartitionStartIndex = 0;
-        for (I interval : intervals) {
-            int[] partitionIndexes = overlapDetector.getOverlaps(interval).stream()
+        Broadcast<OverlapDetector<PartitionLocatable<SimpleInterval>>> overlapDetectorBroadcast = ctx.broadcast(overlapDetector);
+        JavaRDD<PartitionLocatable<I>> indexedIntervals = intervals.map(interval -> {
+            int[] partitionIndexes = overlapDetectorBroadcast.getValue().getOverlaps(interval).stream()
                     .mapToInt(PartitionLocatable::getPartitionIndex).toArray();
             if (partitionIndexes.length == 0) {
-                // interval does not overlap any partition - add it to the last one
-                indexedIntervals.add(new PartitionLocatable<>(lastPartitionStartIndex, interval));
-                continue;
+                final List<SimpleInterval> firstLocatables = firstLocatablesBroadcast.getValue();
+                // interval does not overlap any partition - add it to the one after the interval start
+                int i = Collections.binarySearch(firstLocatables, new SimpleInterval(interval), (o1, o2) -> IntervalUtils.compareLocatables(o1, o2, sequenceDictionary));
+                if (i >= 0) {
+                    throw new IllegalStateException(); // TODO: no overlaps, yet start of interval matches a partition read extent start
+                }
+                int insertionPoint = -i - 1;
```

Comment:
I think this is correct. If the return value is `i`, then `-i - 1` gives the insertion point, since
```
-i - 1 = -((-<insertion point>)-1) - 1
       = <insertion point>
```

E.g. if `i` is -5 then the insertion point is 4. If it's -1 then the insertion point is 0.

Reply:
Ah.  That makes sense.  Just wanted to double-check.